### PR TITLE
Swift API fine tuning and initial documentation of reader API

### PR DESCRIPTION
**Pull request by hassila**
Some start on polish for:
https://github.com/mzaks/FlatBuffersSwift/issues/55
https://github.com/mzaks/FlatBuffersSwift/issues/75

for your consideration - if you are ok with the direction, I will spend some more time on it.

### CODE REVIEW

**Comment by mzaks** [comment]
Yes, I like where it is going!

**Comment by mzaks** [comment]
One thing. I would rather keep all benchmark code in https://github.com/mzaks/FlatBuffersSwiftPerformanceTest
This way it can be more experimental.
So it would be nice if you could delete: `FlatBuffersPerformanceTestDesktop/bench.swift`

**Comment by hassila** [comment]
Sorry, that should have been gone, should be fixed now.

**Comment by mzaks** [comment]
What do you think about dropping the `get` prefix. It is a question of taste I am not sure about it. I think `get` prefix is common in Java but not sure if I saw Swift API using `get` prefix much. E.g 

`try getScalar(at: offset)` 
vs. 
`try scalar(at: offset)` 

or 

`guard index < getVectorElementCount(vectorOffset: vectorOffset) else {` 
vs.
`guard index < vectorCount(at: vectorOffset) else {`

What do you think?

**Comment by hassila** [comment]
I think that would be good actually, I can continue polish in that direction?

**Comment by hassila** [comment]
Ok, will push through a few other changes, can wait with merge of the PR for a while.

**Comment by mzaks** [comment]
OK! Ping me when you think PR is ready to be merged.

**Comment by mzaks** [comment]
@hassila what do you think about renaming `Scalar` to `FlatBuffersScalar` I stumbled upon it because I am also porting FlexBuffer - a schema less binary encoding also done by Wouter von Oortmerssen inf the FlatBuffers project. There I also have a `Scalar` type. Normally I would say it should be solved by Modules, but in case of FlatBuffers Swift and whole module optimisation compiler flag, it might be better to have a good old prefix to distinguish the types. I am not sure.

**Comment by hassila** [comment]
Just a quick question - why do you want to keep both projects together and not separated?

**Comment by mzaks** [comment]
I actually want to keep them separated. But they might end up in the same project as dependencies. `Scalar` is just such a common name that it could collide and because whole module optimisation user might want to keep it as source and not as module dependency

**Comment by hassila** [comment]
Ok, I think it is ready for review and merge if ok - will take a stab at the builder separately. Should probably wait with codegen changes for a bit though as there may be more iterations...

**Comment by hassila** [comment]
"I actually want to keep them separated. But they might end up in the same project as dependencies. Scalar is just such a common name that it could collide and because whole module optimisation user might want to keep it as source and not as module dependency"

Ah, I see. 

Perhaps something could be done with nested types, would need some digging... 

Otherwise the short term fix with `FlatBuffersScalar` would be ok I think, to allow for easy benefit of WMO as you suggest.

**Comment by hassila** [comment]
I'll do an iteration on the builder too now as part of this PR, so please hold again. WIP.

**Comment by mzaks** [comment]
Sorry I didn't merge it yet. I first waiting for travis for a while and than I forgot 😳
I will wait now.

**Comment by hassila** [comment]
No problems, will ping again when next revision is done.

**Comment by hassila** [comment]
Ok, I think it is ready for your review!

File: FlatBuffersSwift/FlatBuffersBuilder.swift [hunk 1]
```diff
@@ -100,69 +126,108 @@ public final class FBBuilder {
...
     }
     
-    public func put<T : Scalar>(value : T){
+    /**
+     Append a scalar value to the buffer
+     
+     - parameters:
+         - value: The value to add to the buffer
+     */
+    public func append<T : Scalar>(value : T){
```
Comment (mzaks) [inline FlatBuffersSwift/FlatBuffersBuilder.swift]: Technically it is a prepend. The buffers are built from back to front. @hassila I leave it up to you, if you want to rename it.

Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Ok, but I would probably argue for that the directionality of the layout is a hidden implementation detail (and probably append would then be easier to read at the call site, as the generated code using the append wouldn't know how the buffer is actually updated?), so I would vote for keeping it as is.
 
We can probably consider additional documentation on the implementation details separately (which would not be a bad thing either :-) )

File: FlatBuffersSwift/FlatBuffersBuilder.swift [hunk 2]
```diff
@@ -100,69 +126,108 @@ public final class FBBuilder {
...
+         - value: The new value 
+         - index: The position to modify
+     */
+    private func update<T : Scalar>(value : T, at index : Int) {
         _data.storeBytes(of: value, toByteOffset: index + leftCursor, as: T.self)
     }
     
-    public func openObject(numOfProperties : Int) throws {
+    /**
+     Start an object update sequence
```
Comment (mzaks) [inline FlatBuffersSwift/FlatBuffersBuilder.swift]: Not sure about update. I guess Start object construction sequence, is a bit more appropriate.

Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Fixed.

File: FlatBuffersSwift/FlatBuffersBuilder.swift [hunk 3]
```diff
@@ -172,51 +237,78 @@ public final class FBBuilder {
...
+     
+     - parameters:
+         - propertyIndex: The index of the property to update
+         - offset: The offsetnumber of properties we will update
+
+     - Returns: The current cursor position (Note: What is the use case of the return value?)
+     */
     @discardableResult
-    public func addPropertyOffsetToOpenObject(propertyIndex : Int, offset : Offset) throws -> Int{
+    public func append(propertyIndex : Int, offset : Offset) throws -> Int{
```
Comment (mzaks) [inline FlatBuffersSwift/FlatBuffersBuilder.swift]: also here `append`feels wrong. maybe `insert(offset:toStartedObjectAt:)` is very long but describes the function in the best possible way I think.

File: FlatBuffersSwift/FlatBuffersBuilder.swift (hunk 1 above)
Comment (mzaks) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: OK

Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Or would you prefer insert?

Comment (mzaks) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Hm ... `insert` sound more neutral. I think I like it :)

File: FlatBuffersSwift/FlatBuffersBuilder.swift (hunk 3 above)
Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Do you want to reverse the arguments?

Looking at usage, current proposal:

```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.append(propertyIndex: 0, value: true, defaultValue: false)
        try! fbb.append(propertyIndex: 1, offset: sOffset)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

would become:

```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.insert(value: true, defaultValue: false, toStartedObjectAt: 0)
        try! fbb.insert(offset: sOffset, toStartedObjectAt: 1)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

Perhaps:
```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.update(value: true, defaultValue: false, toStartedObjectAt: 0)
        try! fbb.update(offset: sOffset, toStartedObjectAt: 1)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

?

File: FlatBuffersSwift/FlatBuffersBuilder.swift (hunk 1 above)
Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: https://github.com/mzaks/FlatBuffersSwift/pull/76/commits/3c3107dd6b2886a0ba3cf802f51399d177005a9c

File: FlatBuffersSwift/FlatBuffersBuilder.swift (hunk 3 above)
Comment (mzaks) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: `update` gives me a feeling that I can exchange values. IMHO `insert` is ok, also I find the API reads fine.

Comment (hassila) [reply FlatBuffersSwift/FlatBuffersBuilder.swift]: Ok, updated in c306411.

**Comment by mzaks** [comment]
Great job!

**Comment by hassila** [comment]
Thanks, I'm sure more can be done, but it is an initial attempt! :)

**Pull request merged**